### PR TITLE
octave: revision for suite-sparse

### PR DESCRIPTION
**Pull request by ilovezfs**
make head compile

### CODE REVIEW

**Comment by ilovezfs** [comment]
CC @schoeps

**Comment by ilovezfs** [comment]
@schoeps The good news is the segfault on El Capitan was not reproducible. The bad news is the segfault on El Capitan was not reproducible.

**Comment by schoeps** [comment]
Was it during the tests? Possibly related to fontconfig?

**Comment by ilovezfs** [comment]
@schoeps https://github.com/Homebrew/homebrew-science/pull/5101#issuecomment-279189341

File: octave.rb [hunk 1]
```diff
@@ -226,10 +229,13 @@ def install
         args << "--with-blas=-L#{Formula["veclibfort"].opt_lib} -lvecLibFort"
       end
     else # OS.linux? and without "openblas"
-      args << "--with-blas=-lblas -llapack"
+       "--with-blas=-lblas -llapack"
```
Comment (schoeps) [inline octave.rb]: @ilovezfs did I do this change? is it correct?

**Comment by ilovezfs** [comment]
```
  libinterp/corefcn/file-io.cc-tst ............................ PASS     80/81  
                                                                  FAIL    1
  libinterp/corefcn/filter.cc-tst ............................. PASS     14/14  
  libinterp/corefcn/find.cc-tst ............................... PASS     16/16  
  libinterp/corefcn/gammainc.cc-tst ........................... PASS      4/4   
  libinterp/corefcn/gcd.cc-tst ................................ PASS      9/9   
  libinterp/corefcn/getrusage.cc-tst .......................... PASS      1/1   
  libinterp/corefcn/givens.cc-tst ............................. PASS      6/6   
  libinterp/corefcn/graphics.cc-tst ...........................Fontconfig warning: line 146: blank doesn't take any effect anymore. please remove it from your fonts.conf
panic: Segmentation fault: 11 -- stopping myself...
/bin/sh: line 1: 44495 Segmentation fault: 11  /bin/sh ../run-octave --norc --silent --no-history /private/tmp/octave-20170211-18579-ubnbl5/octave-4.2.0/test/fntests.m /private/tmp/octave-20170211-18579-ubnbl5/octave-4.2.0/test
make[3]: *** [check-local] Error 139
make[2]: *** [check-am] Error 2
make[1]: *** [check-recursive] Error 1
make: *** [check] Error 2
```

**Comment by schoeps** [comment]
@ilovezfs Thanks. So it's related to fontconfig. one could add a `fc-cache -frv` before the tests. This seems to help. What do you think?

**Comment by ilovezfs** [comment]
That seems fine to me. I don't think it will cause a sandbox violation.

**Comment by ilovezfs** [comment]
@schoeps if it's fixing a real problem, we probably also want to do it in post-install since it won't happen for people pouring the bottles unless it's there too.

**Comment by ilovezfs** [comment]
@schoeps PR refreshed.

**Comment by schoeps** [comment]
@ilovezfs Thanks! Looks perfect. However, we will probably never know whether it was necessary since the bug was not reproducible :(

**Comment by ilovezfs** [comment]
@schoeps well we can always try removing it again at a later date but better safe than sorry I think.

**Comment by schoeps** [comment]
maybe you could add a comment to the corresponding line? ("should fix segmentation faults due to fontconfig")

**Comment by ilovezfs** [comment]
@schoeps yup good idea. I'll do that "in post" (i.e., after CI).

**Comment by ilovezfs** [comment]
@schoeps here's the panic with qt 5.8 and the development versions of pyqt5, qscintilla2, and sip: https://gist.github.com/ilovezfs/2bb0ed41614234eb2ccd3a77b28b7a9c
```
See the file /private/tmp/octave-20170212-35732-14171k2/octave-4.2.0/test/fntests.log for additional details.

Items listed as XFAIL above are known bugs.
Bug report numbers for them may be found in the log file:
/private/tmp/octave-20170212-35732-14171k2/octave-4.2.0/test/fntests.log
Please help improve Octave by contributing fixes for them.

Tests are most often skipped because the features they require
have been disabled.  Features are most often disabled because
they require dependencies that were not present when Octave
was built.  The configure script should have printed a summary
at the end of its run indicating which dependencies were not found.

0 (of 0) .m files have no tests.

Please help improve Octave by contributing tests for these files
(see the list in the file /private/tmp/octave-20170212-35732-14171k2/octave-4.2.0/test/fntests.log).

panic: Abort trap: 6 -- stopping myself...
attempting to save variables to 'octave-workspace'...
```

**Comment by ilovezfs** [comment]
@schoeps build failure on Linux is
```
  CXXLD    src/octave-cli
  GEN      libinterp/dldfcn/__delaunayn__.oct
  GEN      libinterp/dldfcn/__eigs__.oct
  GEN      libinterp/dldfcn/__fltk_uigetfile__.oct
  GEN      libinterp/dldfcn/__glpk__.oct
  GEN      libinterp/dldfcn/__init_fltk__.oct
  GEN      libinterp/dldfcn/__init_gnuplot__.oct
  GEN      libinterp/dldfcn/__osmesa_print__.oct
  GEN      libinterp/dldfcn/__voronoi__.oct
  GEN      libinterp/dldfcn/amd.oct
  GEN      libinterp/dldfcn/audiodevinfo.oct
  GEN      libinterp/dldfcn/audioread.oct
  GEN      libinterp/dldfcn/ccolamd.oct
  GEN      libinterp/dldfcn/chol.oct
  GEN      libinterp/dldfcn/colamd.oct
  GEN      libinterp/dldfcn/convhulln.oct
  GEN      libinterp/dldfcn/dmperm.oct
  GEN      libinterp/dldfcn/fftw.oct
  GEN      libinterp/dldfcn/gzip.oct
  GEN      libinterp/dldfcn/qr.oct
  GEN      libinterp/dldfcn/symbfact.oct
  GEN      libinterp/dldfcn/symrcm.oct
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::ping(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::attribute(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::CoderInfo::description[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::comment[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::magick[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::CoderInfo::CoderInfo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/octave-cli] Error 1
make[2]: Leaving directory `/tmp/octave-20170212-123783-1ixdeve/octave-4.2.0'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/octave-20170212-123783-1ixdeve/octave-4.2.0'
make: *** [all] Error 2
```

**Comment by ilovezfs** [comment]
OK, it built this time, but on circle it failed with
```
==> Summary
/home/linuxbrew/.linuxbrew/Cellar/octave/4.2.0_4: 2,182 files, 45.1M
==> brew test homebrew/science/octave --verbose
Testing homebrew/science/octave
==> /home/linuxbrew/.linuxbrew/Cellar/octave/4.2.0_4/bin/octave --eval (22/7 - pi)/pi
.
Last 150 lines from /home/ubuntu/.cache/Homebrew/Logs/octave/test.01.octave:
2017-02-12 14:18:38 +0000

/home/linuxbrew/.linuxbrew/Cellar/octave/4.2.0_4/bin/octave
--eval
(22/7 - pi)/pi

/home/linuxbrew/.linuxbrew/Cellar/octave/4.2.0_4/bin/octave-cli-4.2.0: symbol lookup error: /home/linuxbrew/.linuxbrew/lib/libX11.so.6: undefined symbol: xcb_wait_for_reply64
Error: homebrew/science/octave: failed
Failed executing: /home/linuxbrew/.linuxbrew/Cellar/octave/4.2.0_4/bin/octave --eval (22/7\ -\ pi)/pi
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1834:in `block in system'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1772:in `open'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1772:in `system'
/home/linuxbrew/.linuxbrew/Library/Taps/homebrew/homebrew-science/octave.rb:347:in `block in <class:Octave>'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1671:in `block (2 levels) in run_test'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:884:in `with_logging'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1670:in `block in run_test'
/home/linuxbrew/.linuxbrew/Library/Homebrew/extend/fileutils.rb:14:in `block in mktemp'
/home/linuxbrew/.linuxbrew/Library/Homebrew/extend/fileutils.rb:74:in `block in run'
/home/linuxbrew/.linuxbrew/Library/Homebrew/extend/fileutils.rb:74:in `chdir'
/home/linuxbrew/.linuxbrew/Library/Homebrew/extend/fileutils.rb:74:in `run'
/home/linuxbrew/.linuxbrew/Library/Homebrew/extend/fileutils.rb:13:in `mktemp'
/home/linuxbrew/.linuxbrew/Library/Homebrew/formula.rb:1664:in `run_test'
/home/linuxbrew/.linuxbrew/Library/Homebrew/test.rb:33:in `block in <main>'
/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/home/linuxbrew/.linuxbrew/Library/Homebrew/test.rb:32:in `<main>'
==> FAILED
```

However on Travis it passed.

**Comment by schoeps** [comment]
Why do you think it's gcc's fault? And do you care? It's ancient :)

P.S.: I am not very helpful today, a presentation has to be prepared for next week.

**Comment by sjackman** [comment]
I've updated the `circle.yml` to remove the host package `libx11-dev`. With any luck, that should hopefully resolve this issue. Rebase this PR on top of master and force push.

**Comment by ilovezfs** [comment]
rebased.

**Comment by sjackman** [comment]
Travis failed while building `pstoedit`. I've opened a PR to bottle it for Linuxbrew at https://github.com/Linuxbrew/homebrew-core/pull/1776
```
latex2man -t pstoedit.trans -M pstoedit.tex pstoedit.1 
/bin/bash: latex2man: command not found
```
https://travis-ci.org/Homebrew/homebrew-science/builds/200875714#L6065

**Comment by ilovezfs** [comment]
Circle failure is
```
  CXX      libinterp/libinterp_liboctinterp_la-build-env-features.lo
  CXXLD    src/mkoctfile
  CXXLD    src/octave
  CXXLD    src/octave-config
  CXXLD    liboctave/liboctave.la
  CXXLD    libinterp/liboctinterp.la
  CXXLD    libinterp/dldfcn/__delaunayn__.la
  CXXLD    libinterp/dldfcn/__eigs__.la
  CXXLD    libinterp/dldfcn/__fltk_uigetfile__.la
  CXXLD    libinterp/dldfcn/__glpk__.la
/usr/bin/ld: cannot find -lfltk_gl
collect2: error: ld returned 1 exit status
make[2]: *** [libinterp/dldfcn/__fltk_uigetfile__.la] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/tmp/octave-20170212-120591-14p2cxm/octave-4.2.0'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/octave-20170212-120591-14p2cxm/octave-4.2.0'
make: *** [all] Error 2
```

**Comment by sjackman** [comment]
Add `depends_on `fltk` unless OS.mac?`

**Comment by ilovezfs** [comment]
@sjackman unfortunately already there. See

https://github.com/ilovezfs/homebrew-science/blob/ce118ec32c1ae80eb82f07b3d08ab193500595c0/octave.rb#L118

and

https://github.com/ilovezfs/homebrew-science/blob/ce118ec32c1ae80eb82f07b3d08ab193500595c0/octave.rb#L201

**Comment by sjackman** [comment]
I've bottled `fltk` and other `octave` dependencies for Linuxbrew.

**Comment by sjackman** [comment]
The `fltk` bottle includes `libfltk_gl.so`. My guess with the previous error was that the `mesa` dependency wasn't installed.

**Comment by ilovezfs** [comment]
@sjackman so other than the qt@5.7 build failing on Linux, the other issue is
```
  GEN      libinterp/dldfcn/symbfact.oct
  GEN      libinterp/dldfcn/symrcm.oct
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::ping(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::attribute(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::CoderInfo::description[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::comment[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::Image::magick[abi:cxx11]() const'
libinterp/.libs/liboctinterp.so: undefined reference to `Magick::CoderInfo::CoderInfo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/octave-cli] Error 1
make[2]: Leaving directory `/tmp/octave-20170212-71735-tnyxs0/octave-4.2.0'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/tmp/octave-20170212-71735-tnyxs0/octave-4.2.0'
make: *** [all] Error 2
```
which I've avoided by making `graphicsmagick` optional instead of recommended on Linux. Any ideas on how to fix that properly?

**Comment by ilovezfs** [comment]
:green_apple: at long last.

**Comment by sjackman** [comment]
Thanks for getting the Linux build to 💚. I appreciate it.

Assuming that `-lgraphicsmagick` is at the right place in the command line, my best guess would be a libstdc++ ABI incompatibility due to incompatible `g++ -std=x`. I'd have to look into it further.

**Pull request closed without merging**